File: legacy/tui/prtui/data/store.py
"""Persistent comment + Claude-session store (JSON, atomic writes, tombstones)."""

from __future__ import annotations

import json
import os
import tempfile
import time
import uuid
from dataclasses import asdict, dataclass, field
from pathlib import Path

SCHEMA_VERSION = 1


def _state_root() -> Path:
    override = os.environ.get("PRTUI_STATE_DIR")
    if override:
        return Path(override)
    base = os.environ.get("XDG_STATE_HOME") or os.path.expanduser("~/.local/state")
    return Path(base) / "prtui"


def _now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def _from_dict(cls, data: dict):
    """Build a dataclass, ignoring unknown keys (tolerates schema drift)."""
    if not isinstance(data, dict):
        return None
    fields = {f.name for f in __import__("dataclasses").fields(cls)}
    try:
        return cls(**{k: v for k, v in data.items() if k in fields})
    except TypeError:
        return None


@dataclass
class Comment:
    id: str
    file: str
    side: str  # LEFT|RIGHT
    line_start: int
    line_end: int
    body: str
    origin: str = "local"  # local|github|claude
    status: str = "draft"  # draft|published|resolved|outdated
    kind: str = "normal"  # normal|suggestion
    suggestion_text: str | None = None
    in_reply_to: str | None = None
    github_id: str | None = None
    author: str = "you"
    created_at: str = field(default_factory=_now)
    updated_at: str = field(default_factory=_now)
    hidden: bool = False


@dataclass
class Session:
    id: str
    state: str = "running"  # running|done|error
    verdict: str | None = None
    summary: str = ""
    instruction: str = ""
    allow_edits: bool = False
    auto_resolve: bool = False
    started_at: str = field(default_factory=_now)
    ended_at: str | None = None
    replied: list[str] = field(default_factory=list)
    findings: list[dict] = field(default_factory=list)
    commits: list[dict] = field(default_factory=list)
    log: list[str] = field(default_factory=list)
    error: str | None = None
    applied: bool = False


class Store:
    def __init__(self, source_key: str, repo_root: str):
        self.source_key = source_key
        self.repo_root = repo_root
        self.comments: dict[str, Comment] = {}
        self.sessions: dict[str, Session] = {}
        self.tombstones: dict[str, str] = {}
        self._load()

    # --- persistence ---------------------------------------------------------
    def _path(self) -> Path:
        import hashlib
        rk = hashlib.blake2s(self.repo_root.encode(), digest_size=4).hexdigest()
        sk = hashlib.blake2s(self.source_key.encode(), digest_size=4).hexdigest()
        return _state_root() / rk / f"{sk}.json"

    def _load(self) -> None:
        p = self._path()
        if not p.exists():
            return
        try:
            doc = json.loads(p.read_text())
        except (json.JSONDecodeError, OSError):
            return
        self.tombstones = doc.get("tombstones", {})
        for cid, c in (doc.get("comments") or {}).items():
            if cid not in self.tombstones:
                obj = _from_dict(Comment, c)
                if obj:
                    self.comments[cid] = obj
        for sid, s in (doc.get("sessions") or {}).items():
            obj = _from_dict(Session, s)
            if obj:
                self.sessions[sid] = obj

    def save(self) -> None:
        p = self._path()
        p.parent.mkdir(parents=True, exist_ok=True)
        # Reload-merge tombstones from disk so concurrent deletes survive.
        disk = {}
        if p.exists():
            try:
                disk = json.loads(p.read_text())
            except (json.JSONDecodeError, OSError):
                disk = {}
        merged_tombstones = {**disk.get("tombstones", {}), **self.tombstones}
        # Merge disk records another process may have written; in-memory wins on
        # conflict, tombstoned ids are dropped.
        comments = {cid: c for cid, c in (disk.get("comments") or {}).items()
                    if cid not in merged_tombstones}
        comments.update({cid: asdict(c) for cid, c in self.comments.items()
                         if cid not in merged_tombstones})
        sessions = dict(disk.get("sessions") or {})
        sessions.update({sid: asdict(s) for sid, s in self.sessions.items()})
        doc = {
            "schema_version": SCHEMA_VERSION,
            "source_key": self.source_key,
            "comments": comments,
            "sessions": sessions,
            "tombstones": merged_tombstones,
        }
        fd, tmp = tempfile.mkstemp(dir=str(p.parent), prefix=".tmp")
        with os.fdopen(fd, "w") as fh:
            json.dump(doc, fh)
        os.replace(tmp, str(p))

    # --- comment CRUD --------------------------------------------------------
    def add(self, file: str, side: str, line_start: int, body: str,
            line_end: int | None = None, origin: str = "local",
            kind: str = "normal", suggestion_text: str | None = None) -> Comment:
        c = Comment(
            id=str(uuid.uuid4()), file=file, side=side,
            line_start=line_start, line_end=line_end or line_start,
            body=body, origin=origin, kind=kind, suggestion_text=suggestion_text,
            author=os.environ.get("USER", "you"),
        )
        self.comments[c.id] = c
        self.save()
        return c

    def reply(self, parent_id: str, body: str, origin: str = "local",
              suggestion_text: str | None = None) -> Comment | None:
        parent = self.comments.get(parent_id)
        if not parent:
            return None
        root = self.root_of(parent_id)
        c = Comment(
            id=str(uuid.uuid4()), file=parent.file, side=parent.side,
            line_start=parent.line_start, line_end=parent.line_end, body=body,
            origin=origin, in_reply_to=root,
            kind="suggestion" if suggestion_text else "normal",
            suggestion_text=suggestion_text,
            author=os.environ.get("USER", "you"),
        )
        self.comments[c.id] = c
        self.save()
        return c

    def root_of(self, cid: str) -> str:
        c = self.comments.get(cid)
        return c.in_reply_to or cid if c else cid

    def delete(self, cid: str) -> bool:
        c = self.comments.get(cid)
        if not c:
            return False
        now = _now()
        if not c.in_reply_to:
            for other_id, other in list(self.comments.items()):
                if other.in_reply_to == cid:
                    del self.comments[other_id]
                    self.tombstones[other_id] = now
        del self.comments[cid]
        self.tombstones[cid] = now
        self.save()
        return True

    def set_resolved(self, cid: str, resolved: bool) -> None:
        root = self.root_of(cid)
        for c in self.comments.values():
            if c.id == root or c.in_reply_to == root:
                c.status = "resolved" if resolved else "draft"
                c.updated_at = _now()
        self.save()

    def threads_for_file(self, file: str) -> list[Comment]:
        roots = [c for c in self.comments.values() if c.file == file and not c.in_reply_to]
        roots.sort(key=lambda c: c.line_start)
        return roots

    def all_threads(self) -> list[Comment]:
        return [c for c in self.comments.values() if not c.in_reply_to]

    def replies(self, root_id: str) -> list[Comment]:
        out = [c for c in self.comments.values() if c.in_reply_to == root_id]
        out.sort(key=lambda c: c.created_at)
        return out

    def get(self, cid: str) -> Comment | None:
        return self.comments.get(cid)

File: lua/review/comments/store.lua
-- review.nvim :: comment store — CRUD, threads, persistence, re-anchoring.
--
-- One Store instance per open Source. Backed by state.lua (merge-on-write).

local state = require("review.state")
local anchor = require("review.comments.anchor")
local util = require("review.util")

---@class Store
---@field source table
---@field repo_key string
---@field source_key string
---@field comments table<string, table>
---@field sessions table<string, table>
local Store = {}
Store.__index = Store

--- Load (or create) the store for a source.
---@param source table
---@return Store
function Store.for_source(source)
  local repo_key = source:repo_key()
  local source_key = source:key()
  local doc = state.load(repo_key, source_key)
  local self = setmetatable({
    source = source,
    repo_key = repo_key,
    source_key = source_key,
    comments = doc.comments or {},
    sessions = doc.sessions or {},
    tombstones = doc.tombstones or {},
    viewed = (doc.meta and doc.meta.viewed) or {},
    review = (doc.meta and doc.meta.review) or { event = "COMMENT", body = "" },
  }, Store)
  -- Defensively drop any tombstoned records that lingered on disk.
  for id in pairs(self.tombstones) do
    self.comments[id] = nil
  end
  return self
end

--- Persist current state (merge-on-write).
function Store:save()
  local ok, err = state.save(self.repo_key, self.source_key, {
    schema_version = state.SCHEMA_VERSION,
    source_key = self.source_key,
    comments = self.comments,
    sessions = self.sessions,
    tombstones = self.tombstones,
    meta = { viewed = self.viewed, review = self.review },
  })
  if not ok then
    util.notify("failed to persist comments: " .. tostring(err), vim.log.levels.ERROR)
  end
  return ok
end

--- The pending review submission (event + summary body). Persisted with the store so
--- a summary drafted over several sittings survives restarts.
---@return table { event=string, body=string }
function Store:review_draft()
  self.review = self.review or { event = "COMMENT", body = "" }
  self.review.event = self.review.event or "COMMENT"
  self.review.body = self.review.body or ""
  return self.review
end

---@param fields table { event?=string, body?=string }
function Store:set_review_draft(fields)
  local draft = self:review_draft()
  for key, value in pairs(fields or {}) do
    draft[key] = value
  end
  self:save()
  return draft
end

function Store:is_viewed(file)
  return self.viewed[file] == self.source:head_rev()
end

function Store:set_viewed(file, value)
  self.viewed[file] = value == false and nil or self.source:head_rev()
  self:save()
  return self:is_viewed(file)
end

function Store:viewed_progress()
  local files, viewed = self.source:files(), 0
  for _, file in ipairs(files) do if self:is_viewed(file.path) then viewed = viewed + 1 end end
  return viewed, #files
end

--- The status a non-resolved, non-outdated comment should carry. Anything with an
--- upstream id is already published; only local-only records are drafts.
---@param c table
---@return string
local function open_status(c)
  return c.github_id and "published" or "draft"
end

--- Add a new root comment.
---@param fields table { file, side, line_start, line_end, body, kind?, suggestion_text?, origin?, head_sha?, base_sha? }
---@return table comment
function Store:add(fields)
  local now = os.date("!%Y-%m-%dT%H:%M:%SZ")
  local meta = self.source:metadata()
  local side = fields.side or "RIGHT"
  local rev = (side == "LEFT") and self.source:base_rev() or self.source:head_rev()
  -- Always anchor against the reviewed revision. The selected branch or commit may
  -- not be checked out in the user's working tree.
  local lines = anchor.file_lines(meta.repo_root, rev, fields.file) or {}
  local a = anchor.compute(lines, fields.line_start)
  if side == "LEFT" then
    a.blob_sha = rev
  end
  local comment = {
    id = util.uuid(),
    source_key = self.source_key,
    file = fields.file,
    side = side,
    head_sha = fields.head_sha or self.source:head_rev(),
    base_sha = fields.base_sha or self.source:base_rev(),
    line_start = fields.line_start,
    line_end = fields.line_end or fields.line_start,
    anchor = a,
    rename_lineage = {},
    kind = fields.kind or "normal",
    suggestion_text = fields.suggestion_text,
    body = fields.body or "",
    origin = fields.origin or "local",
    status = "draft",
    in_reply_to = nil,
    author = fields.author or (fields.origin == "claude" and "claude") or vim.env.USER or "you",
    created_at = now,
    updated_at = now,
    hidden = false,
  }
  self.comments[comment.id] = comment
  self:save()
  return comment
end

--- Reply to an existing comment (creates a child in the same thread/location).
---@param parent_id string
---@param body string
---@param opts table|nil { origin?, suggestion_text?, author? }
---@return table|nil comment, string|nil err
function Store:reply(parent_id, body, opts)
  opts = opts or {}
  local parent = self.comments[parent_id]
  if not parent then
    return nil, "no such comment: " .. tostring(parent_id)
  end
  local now = os.date("!%Y-%m-%dT%H:%M:%SZ")
  local comment = vim.deepcopy(parent)
  comment.id = util.uuid()
  comment.in_reply_to = self:root_of(parent_id)
  comment.body = body
  comment.origin = opts.origin or "local"
  comment.status = "draft"
  comment.suggestion_text = opts.suggestion_text
  comment.kind = opts.suggestion_text and "suggestion" or "normal"
  comment.github_id = nil
  comment.author = opts.author or (opts.origin == "claude" and "claude") or vim.env.USER or "you"
  comment.created_at = now
  comment.updated_at = now
  self.comments[comment.id] = comment
  self:save()
  return comment
end

--- The root id of a thread containing `id`.
---@param id string
---@return string
function Store:root_of(id)
  local c = self.comments[id]
  if not c then
    return id
  end
  return c.in_reply_to or id
end

--- Update fields of a comment.
---@param id string
---@param fields table
---@return boolean
function Store:update(id, fields)
  local c = self.comments[id]
  if not c then
    return false
  end
  for k, v in pairs(fields) do
    c[k] = v
  end
  c.updated_at = os.date("!%Y-%m-%dT%H:%M:%SZ")
  self:save()
  return true
end

--- Delete a comment (and its direct replies if it's a root).
---@param id string
---@return boolean
function Store:delete(id)
  local c = self.comments[id]
  if not c then
    return false
  end
  local now = os.date("!%Y-%m-%dT%H:%M:%SZ")
  if not c.in_reply_to then
    -- root: remove whole thread (tombstone each so deletes survive merge)
    for cid, other in pairs(self.comments) do
      if other.in_reply_to == id then
        self.comments[cid] = nil
        self.tombstones[cid] = now
      end
    end
  end
  self.comments[id] = nil
  self.tombstones[id] = now
  self:save()
  return true
end

--- Toggle/resolve a thread by root id.
---@param id string
---@param resolved boolean
function Store:set_resolved(id, resolved)
  local root = self:root_of(id)
  for _, c in pairs(self.comments) do
    if c.id == root or c.in_reply_to == root then
      c.status = resolved and "resolved" or open_status(c)
      c.updated_at = os.date("!%Y-%m-%dT%H:%M:%SZ")
    end
  end
  self:save()
end

--- Hide/unhide a thread.
---@param id string
---@param hidden boolean
function Store:set_hidden(id, hidden)
  local root = self:root_of(id)
  for _, c in pairs(self.comments) do
    if c.id == root or c.in_reply_to == root then
      c.hidden = hidden
    end
  end
  self:save()
end

--- All root comments (threads) for a file, sorted by line.
---@param file string
---@return table[] roots
function Store:threads_for_file(file)
  local roots = {}
  for _, c in pairs(self.comments) do
    if c.file == file and not c.in_reply_to then
      table.insert(roots, c)
    end
  end
  table.sort(roots, function(a, b)
    return (a.line_start or 0) < (b.line_start or 0)
  end)
  return roots
end

--- Replies for a root id, chronological.
---@param root_id string
---@return table[]
function Store:replies(root_id)
  local out = {}
  for _, c in pairs(self.comments) do
    if c.in_reply_to == root_id then
      table.insert(out, c)
    end
  end
  table.sort(out, function(a, b)
    return (a.created_at or "") < (b.created_at or "")
  end)
  return out
end

--- All root comments across the source.
---@return table[]
function Store:all_threads()
  local roots = {}
  for _, c in pairs(self.comments) do
    if not c.in_reply_to then
      table.insert(roots, c)
    end
  end
  return roots
end

function Store:get(id)
  return self.comments[id]
end

--- Re-anchor every comment against current file content, applying rename map.
--- Marks comments "outdated" when their anchor can't be uniquely relocated.
---@param rename_map table<string,string>|nil  old_path -> new_path
function Store:reanchor(rename_map)
  local meta = self.source:metadata()
  rename_map = rename_map or {}
  for _, c in pairs(self.comments) do
    -- Only re-anchor root comments; replies inherit the root's location for display
    -- (markers/side-panel key off roots), so relocating them independently is wasted
    -- work and would let a reply's line drift from its root.
    if c.status ~= "resolved" and not c.in_reply_to then
      -- Apply rename lineage — RIGHT side only. LEFT anchors point at the base blob
      -- under the OLD path, so renaming their file would break `git show base:path`.
      if c.side ~= "LEFT" then
        local newp = rename_map[c.file]
        if newp and newp ~= c.file then
          c.rename_lineage = c.rename_lineage or {}
          table.insert(c.rename_lineage, c.file)
          c.file = newp
        end
      end
      local rev = (c.side == "LEFT") and (c.anchor and c.anchor.blob_sha or c.base_sha) or nil
      local lines = anchor.file_lines(meta.repo_root, rev, c.file)
      if lines then
        local loc = anchor.relocate(lines, c.anchor)
        if loc then
          local delta = loc - c.line_start
          c.line_start = loc
          c.line_end = c.line_end + delta
          if c.status == "outdated" then
            c.status = open_status(c)
          end
        else
          c.status = "outdated"
        end
      end
    end
  end
  self:save()
end

return Store
